### PR TITLE
memory-game

### PR DESCRIPTION
**Pull request by KonstantinOkhorzin**
### Memory – Pair Game

[Demo](https://konstantinokhorzin.github.io/memory-game/) |
[Code base](https://github.com/KonstantinOkhorzin/memory-game)

The code is submitted in a dedicated feature branch.

Only code files are submitted.

Please, review.

### CODE REVIEW

**Comment by artem-trubin** [comment]
Hi! There is an easily reproducible bug: if you click a lot of cards fast it just starts opening them all and you can win by clicking all of them. Please fix it before further code review.

**Comment by KonstantinOkhorzin** [comment]
I fixed this.bug. Please, review.

File: submissions/KostiantynOkhorzin/memory-game/css/style.css [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+  border: 0;
+}
+
+h1, h2, h3, h4, h5, h6 {
+  font-size: inherit;
+  font-weight: inherit;
+}
+
+* {
+  font-family: Arial, Helvetica, sans-serif;
```
Comment (artem-trubin) [inline submissions/KostiantynOkhorzin/memory-game/css/style.css:42]: You already establish styles for `*`, why not do it there?

File: submissions/KostiantynOkhorzin/memory-game/script.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    { imgSrc: "img/lime.png", name: "lime" },
+    { imgSrc: "img/ananas.png", name: "ananas" },
+    { imgSrc: "img/apple.png", name: "apple" },
+    { imgSrc: "img/avocado.png", name: "avocado" },
+    { imgSrc: "img/banana.png", name: "banana" },
+    { imgSrc: "img/kiwi.png", name: "kiwi" },
+    { imgSrc: "img/lime.png", name: "lime" },
+];
+
+const styleList = {
```
Comment (artem-trubin) [inline submissions/KostiantynOkhorzin/memory-game/script.js:22]: What is the purpose of this object? You don't use it in the `render()` method of a `GameCard` on line 35.

File: submissions/KostiantynOkhorzin/memory-game/script.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        e.target.classList.add(styleList.selectedCard, styleList.flippedCard);
+        const flippedCards = document.querySelectorAll(`.${styleList.flippedCard}`);
+        const selectedCards = document.querySelectorAll(`.${styleList.selectedCard}`);
+        if (selectedCards.length === 2) {
+            disableCards();
+            if (selectedCards[0].dataset.name === selectedCards[1].dataset.name) {
+                selectedCards.forEach((card) => {
+                    card.classList.remove(styleList.selectedCard);
+                    setTimeout(() => card.classList.add(styleList.hiddenCard), time);
+                })
```
Comment (artem-trubin) [inline submissions/KostiantynOkhorzin/memory-game/script.js:74]: Why does flipping a correct pair of cards doesn't count as a `playerStep`?

File: submissions/KostiantynOkhorzin/memory-game/script.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        playerSteps = 0;
+        renderCards();
+    }, 2000);
+};
+
+renderCards();
+
+gameBody.addEventListener('click', flippingCards);
+
+
```
Comment (artem-trubin) [inline submissions/KostiantynOkhorzin/memory-game/script.js:102]: Two empty lines

File: submissions/KostiantynOkhorzin/memory-game/script.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    const cards = document.querySelectorAll('.game__card');
+    cards.forEach(card => {
+        card.setAttribute('disabled', '');
+        setTimeout(() => {
+            card.removeAttribute('disabled');
+        }, 3000);
+    });
+};
+
+const flippingCards = (e) => {
```
Comment (artem-trubin) [inline submissions/KostiantynOkhorzin/memory-game/script.js:63]: Strange name. I believe something like `handleCardFlip` should be more appropriate.

**Review by artem-trubin — CHANGES_REQUESTED** [review]
Hi! Nice work! Let's make it even better! Check comments below.

**Comment by KonstantinOkhorzin** [comment]
I fixed all. Please, check the changes.

File: submissions/KostiantynOkhorzin/memory-game/css/style.css [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+@media (max-width: 768px) {
+  .game__body {
+    grid-template-columns: repeat(4, 100px);
+    grid-template-rows: repeat(3, 100px);
+  }
+}
+
+@media (max-width: 425px) {
```
Comment (artem-trubin) [inline submissions/KostiantynOkhorzin/memory-game/css/style.css:76]: You already have a breakpoint of `425px` on the line 56. Just place this code there, no need to create separate `@media` blocks for the same width.

File: submissions/KostiantynOkhorzin/memory-game/script.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
+'use strict';
+
+const gameBody = document.querySelector('.game__body');
+let playerSteps = 0;
+const cardProcessingTime = 1500;
+
+const dataImgs = [
```
Comment (artem-trubin) [inline submissions/KostiantynOkhorzin/memory-game/script.js:7]: You store each item twice here. It doesn't make any sense and should be done in JS.

File: submissions/KostiantynOkhorzin/memory-game/script.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+const randomlyMixArrayElem = () => {
+    return dataImgs.sort(() => Math.random() - 0.5);
+};
+
+const renderCards = () => {
+    randomlyMixArrayElem().map(({ imgSrc, name }) => new GameCard(imgSrc, name).render());
+};
+
+const handleCardFlip = ({ target }) => {
+    if (target.classList.contains(styleList.card)) {
```
Comment (artem-trubin) [inline submissions/KostiantynOkhorzin/memory-game/script.js:56]: When you see such a situation where you have a whole function body wrapped in one `if`, you can invert its condition and return if that happens instead. That allows you to remove the extra indentation for the whole function and make it more readable. Like this:

```js
if (!target.classList.contains(styleList.card)) return
// ...
```

File: submissions/KostiantynOkhorzin/memory-game/script.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        target.classList.add(styleList.selectedCard, styleList.flippedCard);
+        const flippedCards = document.querySelectorAll(`.${styleList.flippedCard}`);
+        const selectedCards = document.querySelectorAll(`.${styleList.selectedCard}`);
+        if (selectedCards.length === 2) {
+            if (selectedCards[0].dataset.name === selectedCards[1].dataset.name) {
```
Comment (artem-trubin) [inline submissions/KostiantynOkhorzin/memory-game/script.js:61]: This block (lines 61-74) has some repeating code. Try to figure out how you can shorten it.

**Review by artem-trubin — CHANGES_REQUESTED** [review]
Great changes. There are some more comments below, please check them.

File: submissions/KostiantynOkhorzin/memory-game/script.js (hunk 7 above)
Comment (artem-trubin) [reply submissions/KostiantynOkhorzin/memory-game/script.js:7]: You should only store unique items here and double them in your code.

**Comment by KonstantinOkhorzin** [comment]
I fixed all. Please, check the changes.

**Review by artem-trubin — APPROVED** [review]
Good job!

**Pull request merged**